src/ndarray/ndarray_function.cu

```diff
@@ -88,5 +94,110 @@ void Copy<gpu, gpu>(const TBlob &from, TBlob *to,
                         s->stream_);
   }
 }
+
+/*!
+ * \brief GPU impl of elemwise sum for rowsparse tensors.
+ */
+void ElementwiseSumRspImpl(mshadow::Stream<gpu>* s,
+                       const std::vector<NDArray>& nds,
+                       NDArray* out) {
+  using namespace mxnet::op;
+  using namespace rowsparse;
+  using nnvm::dim_t;
+  CHECK_EQ(out->storage_type(), kRowSparseStorage)
+    << "Expected rowsparse storage_type (" << out->storage_type() << " given)";
+  int init = 0;
+  for (const auto& nd : nds) {
+    if (nd.storage_initialized()) {
+      init++;
+    }
+  }
+  if (init == 0) {
+    FillZerosRspImpl<gpu>(s, out);
+    return;
+  }
+  const dim_t num_rows = out->shape()[0];
+  const dim_t row_length = out->shape().ProdShape(1, out->shape().ndim());
+  // TODO(stefan): use temporary workspace from OpContext instead of cudaMalloc
+  MSHADOW_TYPE_SWITCH(out->dtype(), DType, {  // data type
+    MSHADOW_IDX_TYPE_SWITCH(out->aux_type(kIdx), IType, {  // row_idx type
+      // Allocate temporary array row_flg
+      IType* row_flg = NULL;
+      CUDA_CALL(cudaMalloc(&row_flg, num_rows*sizeof(IType)));
```

Comment:
Shortly after, a prefix sum is computed over the row_flg array to compute index locations so that the CUDA kernel are parallelized. For that I would need at least a 32bit integer array. We could consider taking 32bit instead of 64bit but first we would need to make sure that the number of non-zero rows will never go beyond the 32bit integer limit.